File: backend/tournament/points.py
"""Punkte-Berechnung fuer globale Rangliste."""
from __future__ import annotations

from datetime import datetime, timezone

import aiosqlite

PLACEMENT_POINTS = {1: 10, 2: 6, 3: 3, 4: 3}
PARTICIPATION_POINTS = 1
WIN_POINTS_PER_MATCH = 0.5


async def recalculate_player_points(
    db: aiosqlite.Connection,
    tournament_id: int,
) -> None:
    """Berechnet Punkte fuer alle Teilnehmer eines abgeschlossenen Turniers neu."""
    cursor = await db.execute(
        "SELECT tm.discord_id, t.id as team_id FROM team_members tm "
        "JOIN teams t ON tm.team_id = t.id WHERE t.tournament_id = ?",
        (tournament_id,),
    )
    participants = await cursor.fetchall()

    cursor = await db.execute(
        "SELECT team1_id, team2_id, winner_id, round FROM bracket_matches "
        "WHERE tournament_id = ? AND status = 'completed' ORDER BY round DESC",
        (tournament_id,),
    )
    bracket_matches = await cursor.fetchall()

    team_placements: dict[int, int] = {}
    if bracket_matches:
        max_round = max(match["round"] for match in bracket_matches)
        final = next(
            (match for match in bracket_matches if match["round"] == max_round),
            None,
        )
        if final and final["winner_id"]:
            loser_id = (
                final["team1_id"]
                if final["winner_id"] == final["team2_id"]
                else final["team2_id"]
            )
            team_placements[final["winner_id"]] = 1
            if loser_id:
                team_placements[loser_id] = 2
            semifinals = [
                match for match in bracket_matches if match["round"] == max_round - 1
            ]
            for semifinal in semifinals:
                if semifinal["winner_id"]:
                    semifinal_loser = (
                        semifinal["team1_id"]
                        if semifinal["winner_id"] == semifinal["team2_id"]
                        else semifinal["team2_id"]
                    )
                    if semifinal_loser and semifinal_loser not in team_placements:
                        team_placements[semifinal_loser] = 3

    cursor = await db.execute(
        "SELECT bm.winner_id, bm.team1_id, bm.team2_id "
        "FROM bracket_matches bm WHERE bm.tournament_id = ? AND bm.status = 'completed'",
        (tournament_id,),
    )
    bm_rows = await cursor.fetchall()
    team_wins: dict[int, int] = {}
    for bracket_match in bm_rows:
        if bracket_match["winner_id"]:
            winner_id = bracket_match["winner_id"]
            team_wins[winner_id] = team_wins.get(winner_id, 0) + 1

    now = datetime.now(timezone.utc).isoformat()
    for participant in participants:
        discord_id = participant["discord_id"]
        team_id = participant["team_id"]
        placement = team_placements.get(team_id)
        wins = team_wins.get(team_id, 0)

        pts = PARTICIPATION_POINTS
        pts += PLACEMENT_POINTS.get(placement, 0) if placement else 0
        pts += int(wins * WIN_POINTS_PER_MATCH)

        cursor = await db.execute(
            "SELECT * FROM player_points WHERE discord_id = ?",
            (discord_id,),
        )
        existing = await cursor.fetchone()

        if existing:
            new_total = existing["total_points"] + pts
            new_tournaments = existing["tournaments_played"] + 1
            new_matches = existing["matches_played"] + len(bm_rows)
            new_wins = existing["matches_won"] + wins
            new_best = (
                min(placement, existing["best_placement"])
                if placement and existing["best_placement"]
                else (placement or existing["best_placement"])
            )
            await db.execute(
                "UPDATE player_points SET total_points=?, tournaments_played=?, "
                "matches_played=?, matches_won=?, best_placement=?, updated_at=? "
                "WHERE discord_id=?",
                (
                    new_total,
                    new_tournaments,
                    new_matches,
                    new_wins,
                    new_best,
                    now,
                    discord_id,
                ),
            )
        else:
            await db.execute(
                "INSERT INTO player_points (discord_id, total_points, tournaments_played, matches_played, matches_won, best_placement, updated_at) "
                "VALUES (?, ?, ?, ?, ?, ?, ?)",
                (discord_id, pts, 1, len(bm_rows), wins, placement, now),
            )

File: rust/crates/turnier-engine/src/persist/points.rs
//! Punkte-Recompute für die globale Rangliste (Persistenz).
//!
//! Idempotenz-Strategie (bewusste strukturelle Abweichung vom Original, siehe
//! algorithm_notes): Statt additiv in bestehende `player_points`-Zeilen zu
//! akkumulieren — was das Original bei Mehrfach-`completed` doppelt zählte —
//! rechnet diese Funktion die GANZE `player_points`-Tabelle deterministisch aus
//! ALLEN abgeschlossenen, nicht ausgeschlossenen Turnieren neu auf.
//!
//! Für den im Original üblichen Fall (genau ein gerade abgeschlossenes Turnier)
//! sind die resultierenden Zahlen identisch zum additiven Lauf; mehrere
//! Turniere summieren sich genauso, aber ein erneuter Aufruf ändert nichts mehr
//! (idempotent). Die Punkte-/Platzierungs-WERTE je Turnier stammen 1:1 aus
//! [`crate::points`] (inkl. der bug-preserved SE-Round-Heuristik und des
//! globalen `matches_played`).

use std::collections::HashMap;

use sqlx::{Pool, Postgres, Transaction};
use turnier_core::now_utc;

use crate::error::TournamentResult;
use crate::points::{
    contribution_for_team, team_placements, team_wins, CompletedMatch, PointsContribution,
};

// ASCII tag "trnpnts1" as a signed i64. Serializes the global player_points
// DELETE+INSERT recompute across all tournaments for one transaction.
const POINTS_RECOMPUTE_ADVISORY_LOCK_KEY: i64 = i64::from_be_bytes(*b"trnpnts1");

#[derive(sqlx::FromRow)]
struct MatchRow {
    round: i64,
    team1_id: Option<i64>,
    team2_id: Option<i64>,
    winner_id: Option<i64>,
}

/// Aggregierter Stand eines Spielers über alle gewerteten Turniere.
#[derive(Default)]
struct PlayerAggregate {
    total_points: i64,
    tournaments_played: i64,
    matches_played: i64,
    matches_won: i64,
    best_placement: Option<i64>,
}

impl PlayerAggregate {
    fn add(&mut self, c: PointsContribution) {
        self.total_points += c.points;
        self.tournaments_played += 1;
        self.matches_played += c.matches_played;
        self.matches_won += c.matches_won;
        self.best_placement = match (self.best_placement, c.placement) {
            (Some(a), Some(b)) => Some(a.min(b)),
            (Some(a), None) => Some(a),
            (None, b) => b,
        };
    }
}

/// Rechnet die globale Rangliste neu auf. Das Argument `tournament_id` bleibt aus
/// Signatur-Parität erhalten; der Recompute selbst ist turnier-übergreifend und
/// damit idempotent. Eigene Transaktion.
pub async fn recalculate_player_points(
    pool: &Pool<Postgres>,
    tournament_id: i64,
) -> TournamentResult<()> {
    let mut tx = pool.begin().await?;
    recalculate_player_points_in_tx(&mut tx, tournament_id).await?;
    tx.commit().await?;
    Ok(())
}

/// Transaktionsfähige Variante von [`recalculate_player_points`].
///
/// Der Aufrufer besitzt Commit/Rollback. So können Statuswechsel, Audit und
/// Punkte-Recompute wieder wie im Python-Original atomar zusammenlaufen.
pub async fn recalculate_player_points_in_tx(
    tx: &mut Transaction<'_, Postgres>,
    _tournament_id: i64,
) -> TournamentResult<()> {
    sqlx::query("SELECT pg_advisory_xact_lock($1)")
        .bind(POINTS_RECOMPUTE_ADVISORY_LOCK_KEY)
        .execute(&mut **tx)
        .await?;

    // Alle gewerteten Turniere: abgeschlossen UND nicht von der Rangliste
    // ausgeschlossen (der Original-Caller filtert exclude_from_leaderboard vor
    // dem Recalc; hier zentral).
    let tournaments: Vec<(i64,)> = sqlx::query_as(
        "SELECT id FROM turnier.tournaments \
         WHERE status = 'completed' AND exclude_from_leaderboard = false \
         ORDER BY id",
    )
    .fetch_all(&mut **tx)
    .await?;

    let mut aggregates: HashMap<i64, PlayerAggregate> = HashMap::new();

    for (tid,) in &tournaments {
        // Abgeschlossene Bracket-Matches, round DESC (für die Platzierungs-Heuristik).
        let raw_matches: Vec<MatchRow> = sqlx::query_as::<_, MatchRow>(
            "SELECT round, team1_id, team2_id, winner_id FROM turnier.bracket_matches \
             WHERE tournament_id = $1 AND status = 'completed' ORDER BY round DESC, id",
        )
        .bind(tid)
        .fetch_all(&mut **tx)
        .await?;
        let matches: Vec<CompletedMatch> = raw_matches
            .iter()
            .map(|r| CompletedMatch {
                round: r.round,
                team1_id: r.team1_id,
                team2_id: r.team2_id,
                winner_id: r.winner_id,
            })
            .collect();

        let placements = team_placements(&matches);
        let wins = team_wins(&matches);
        let total_completed = matches.len() as i64;

        // Teilnehmer: Mitglieder aller Teams dieses Turniers.
        let participants: Vec<(i64, i64)> = sqlx::query_as(
            "SELECT tm.discord_id, t.id FROM turnier.team_members tm \
             JOIN turnier.teams t ON tm.team_id = t.id \
             WHERE t.tournament_id = $1 ORDER BY t.id, tm.joined_at, tm.id",
        )
        .bind(tid)
        .fetch_all(&mut **tx)
        .await?;

        for (discord_id, team_id) in participants {
            let contribution = contribution_for_team(team_id, &placements, &wins, total_completed);
            aggregates.entry(discord_id).or_default().add(contribution);
        }
    }

    // Tabelle vollständig neu schreiben (idempotent).
    let now = now_utc();
    sqlx::query("DELETE FROM turnier.player_points")
        .execute(&mut **tx)
        .await?;
    for (discord_id, agg) in &aggregates {
        sqlx::query(
            "INSERT INTO turnier.player_points \
             (discord_id, total_points, tournaments_played, matches_played, matches_won, best_placement, updated_at) \
             VALUES ($1, $2, $3, $4, $5, $6, $7)",
        )
        .bind(discord_id)
        .bind(agg.total_points)
        .bind(agg.tournaments_played)
        .bind(agg.matches_played)
        .bind(agg.matches_won)
        .bind(agg.best_placement)
        .bind(now)
        .execute(&mut **tx)
        .await?;
    }

    Ok(())
}
